creusot/src/run_why3.rs

```diff
@@ -0,0 +1,358 @@
+            Some(why3::declaration::Attribute::Span(
+                "rustc_span".into(),
+                data.lo.0 as usize,
+                data.hi.0 as usize,
+                self.encode_span_data((data.ctxt, data.parent)),
+                0
+            ))
+        } else {
+            None
+        }
+    }
+
+    fn decode_span(&self, loc: &Loc) -> Option<Span> {
+        match loc {
+            Loc::Span(Why3Span { file_name, start_line, start_char, end_line, .. })
+                if file_name == "rustc_span" =>
+            {
+                let data = self.decode_span_data(*end_line as usize);
+                Some(Span::new(BytePos(*start_line), BytePos(*start_char), data.0, data.1))
+            }
+            _ => None,
+        }
+    }
+}
+
+fn exp(kind: ExprKind) -> Expr {
+    Expr { id: DUMMY_NODE_ID, kind, span: DUMMY_SP, attrs: Default::default(), tokens: None }
+}
+
+fn pat(kind: PatKind) -> Pat {
+    Pat { id: DUMMY_NODE_ID, kind, span: DUMMY_SP, tokens: None }
+}
+
+fn ty(kind: TyKind) -> Ty {
+    Ty { id: DUMMY_NODE_ID, kind, span: DUMMY_SP, tokens: None }
+}
+
+fn ident(name: &str) -> Ident {
+    Ident { name: Symbol::intern(name), span: DUMMY_SP }
+}
+
+fn name_to_path(name: &str) -> Expr {
+    let segments = name.split(".").into_iter().map(|x| PathSegment::from_ident(ident(x))).collect();
+    exp(ExprKind::Path(None, rustc_ast::Path { span: DUMMY_SP, segments, tokens: None }))
+}
+
+fn lit(name: &str, kind: LitKind, suffix: Option<&str>) -> Expr {
+    exp(ExprKind::Lit(Lit {
+        kind,
+        symbol: Symbol::intern(name),
+        suffix: suffix.map(Symbol::intern),
+    }))
+}
+
+fn fun<'a>(args: impl IntoIterator<Item = &'a str>, body: Expr) -> Expr {
+    exp(ExprKind::Closure(Box::new(rustc_ast::Closure {
+        binder: rustc_ast::ClosureBinder::NotPresent,
+        capture_clause: rustc_ast::CaptureBy::Ref,
+        constness: rustc_ast::Const::No,
+        asyncness: rustc_ast::Async::No,
+        movability: rustc_ast::Movability::Movable,
+        fn_decl: P(rustc_ast::FnDecl {
+            inputs: args
+                .into_iter()
+                .map(|x| rustc_ast::Param {
+                    attrs: Default::default(),
+                    ty: P(ty(TyKind::Infer)),
+                    pat: P(pat(PatKind::Ident(rustc_ast::BindingAnnotation::NONE, ident(x), None))),
+                    id: DUMMY_NODE_ID,
+                    span: DUMMY_SP,
+                    is_placeholder: false,
+                })
+                .collect(),
+            output: rustc_ast::FnRetTy::Default(DUMMY_SP),
+        }),
+        body: P(body),
+        fn_decl_span: DUMMY_SP,
+        fn_arg_span: DUMMY_SP,
+    })))
+}
+
+fn app(f: &str, args: impl IntoIterator<Item = Expr>) -> Expr {
+    let mut v = args.into_iter().map(|x| P(x)).collect();
+    if false {
+        // This is necessary for type checking since ThinVec is not nameable
+        return exp(ExprKind::Call(P(name_to_path(f)), v));
+    }
```

Comment:
? I don't understand this?

Reply:
If I didn't add this, Rust wasn't able to infer the type for
```rust
    let mut v = args.into_iter().map(|x| P(x)).collect();
```
The type is supposed to be `ThinVec` but it isn't `pub use`d in `rustc` so I couldn't just do `let mut v: ThinVec<_> = ... `